numaprom/udf/inference.py

```diff
@@ -35,71 +46,74 @@ def _run_model(
-            payload.uuid,
-            train_payload,
-        )
-        return [("train", orjson.dumps(train_payload))]
 
+def _get_model(payload: StreamPayload, model_config: dict) -> Optional[ArtifactData]:
+    print(model_config)
     artifact_data = load_model(
         skeys=[payload.composite_keys["namespace"], payload.composite_keys["name"]],
         dkeys=[model_config["model_name"]],
     )
-
     if not artifact_data:
+        payload.set_status(Status.ARTIFACT_NOT_FOUND)
         _LOGGER.info(
-            "%s - No model found, sending to trainer. Trainer payload: %s",
+            "%s - Model not found for %s",
             payload.uuid,
-            train_payload,
+            payload.composite_keys,
         )
-        return [("train", orjson.dumps(train_payload))]
-
+        return None
     _LOGGER.debug("%s - Successfully loaded model from mlflow", payload.uuid)
+    return artifact_data
 
-    messages = []
 
+def _is_model_stale(
+    payload: StreamPayload, artifact_data: ArtifactData, model_config: dict
+) -> bool:
     date_updated = artifact_data.extras["last_updated_timestamp"] / 1000
     stale_date = (
         datetime.now() - timedelta(hours=int(model_config["retrain_freq_hr"]))
     ).timestamp()
-
     if date_updated < stale_date:
-        train_payload["resume_training"] = True
         _LOGGER.info(
-            "%s - Model found is stale, sending to trainer. Trainer payload: %s",
+            "%s - Model found is stale for %s",
             payload.uuid,
-            train_payload,
+            payload.composite_keys,
         )
-        messages.append(("train", orjson.dumps(train_payload)))
+        return True
+    return False
+
+
+@conditional_forward
+def inference(_: str, datum: Datum) -> list[tuple[str, bytes]]:
+    _start_time = time.perf_counter()
+
+    _in_msg = datum.value.decode("utf-8")
+    payload = StreamPayload(**orjson.loads(_in_msg))
+
+    messages = []
+
+    # Load config
+    metric_config = get_metric_config(payload.composite_keys["name"])
+    model_config = metric_config["model_config"]
+
+    # Check if model exists
+    artifact_data = _get_model(payload, model_config)
+    if not artifact_data:
+        train_payload = _construct_train_payload(payload, model_config)
+        messages.append((_TRAIN_VTX_KEY, orjson.dumps(train_payload)))
+        return messages
+
+    # Check if current model is stale
+    if _is_model_stale(payload, artifact_data, model_config):
+        train_payload = _construct_train_payload(payload, model_config)
+        messages.append((_TRAIN_VTX_KEY, orjson.dumps(train_payload)))
 
-    messages.append(_run_model(payload, artifact_data, model_config))
+    # Generate predictions
+    payload = _run_inference(payload, artifact_data, model_config)
+    messages.append((_THRESHOLD_VTX_KEY, orjson.dumps(payload, option=orjson.OPT_SERIALIZE_NUMPY)))
 
-    _LOGGER.info("%s - Sending Messages: %s ", payload.uuid, messages)
```

Comment:
We should still log the Messages we are sending, may be make it a debug log or add a log for train_payload too.

Reply:
Yeah sure, adding that